### PR TITLE
Feature/add SurplusBeneficiary

### PR DESCRIPTION
**Pull request by 42bchen**
### PR Reminders

Be ware of the followings

- [ ] update change log
 
---

1. **Contract**: make sure the code follows our convention, ex: naming, explicit returns, etc.; if uncertain, discuss with others

2. **Test**: make sure tests can cover most normal and edge cases; if not, open follow-up tickets. Also, look out for failed tests and fix them!

3. **CHANGELOG.md**: update when external interfaces are changed

4. Workflow: 
    - Github: assign the pr to yourself; if pairing, can merge directly; else, assign someone to help review
    - Asana: assign the corresponding ticket to yourself and leave the pr link on it for easier follow-ups
    - Discord: if someone is mentioned in this pr, tag on Discord

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202448789464186

### CODE REVIEW

File: contracts/SurplusBeneficiary.sol [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import { SafeMathUpgradeable } from "@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
+import { FullMath } from "@uniswap/v3-core/contracts/libraries/FullMath.sol";
+import { SurplusBeneficiaryStorageV1 } from "./storage/SurplusBeneficiaryStorage.sol";
+import { ISurplusBeneficiary } from "./interface/ISurplusBeneficiary.sol";
+import { IFeeDistributor } from "./interface/IFeeDistributor.sol";
+
+contract SurplusBeneficiary is
+    ISurplusBeneficiary,
+    ReentrancyGuardUpgradeable,
+    OwnableUpgradeable,
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:22]: any counter argument for making it immutable?

File: contracts/SurplusBeneficiary.sol [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        setTreasuryPercentage(treasuryPercentageArg);
+
+        IERC20Upgradeable(_token).approve(feeDistributorArg, uint256(-1));
+    }
+
+    //
+    // PUBLIC ONLY-ADMIN NON-VIEW
+    //
+
+    function setToken(address tokenArg) public onlyOwner {
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:51]: i think we should not change token. if we want to change token, we should deploy another instance

File: contracts/SurplusBeneficiary.sol [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    function setTreasury(address treasuryArg) public onlyOwner {
+        // SB_TNC: treasury address is not contract
+        require(treasuryArg.isContract(), "SB_TNC");
+        emit TreasuryChanged(_treasury, treasuryArg);
+        _treasury = treasuryArg;
+    }
+
+    function setTreasuryPercentage(uint24 treasuryPercentageArg) public onlyOwner {
+        // SB_TPZ: treasury percentage is equal to zero
+        require(treasuryPercentageArg != 0, "SB_TPZ");
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:74]: why can't be 0?
should check it's <= 1e6

File: contracts/SurplusBeneficiary.sol [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    //
+    // PUBLIC ONLY-ADMIN NON-VIEW
+    //
+
+    function setToken(address tokenArg) public onlyOwner {
+        // SB_TANC: token address is not contract
+        require(tokenArg.isContract(), "SB_TANC");
+        emit TokenChanged(_token, tokenArg);
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:54]: why do we need to emit old value?

File: contracts/SurplusBeneficiary.sol [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    function setFeeDistributor(address feeDistributorArg) public onlyOwner {
+        // SB_FDNC: feeDistributor address is not contract
+        require(feeDistributorArg.isContract(), "SB_FDNC");
+        emit FeeDistributorChanged(_feeDistributor, feeDistributorArg);
+        _feeDistributor = feeDistributorArg;
+    }
+
+    function setTreasury(address treasuryArg) public onlyOwner {
+        // SB_TNC: treasury address is not contract
+        require(treasuryArg.isContract(), "SB_TNC");
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:67]: why treasury must be a contract?

File: contracts/SurplusBeneficiary.sol [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    /// @inheritdoc ISurplusBeneficiary
+    function dispatch() external override nonReentrant returns (uint256) {
+        address token = _token;
+
+        uint256 tokenAmount = IERC20Upgradeable(token).balanceOf(address(this));
+
+        // SB_TAZ: token amount is zero
+        require(tokenAmount != 0, "SB_TAZ");
+
+        uint256 tokenAmountToTreasury = FullMath.mulDiv(tokenAmount, _treasuryPercentage, 1e6);
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:92]: suggest to add a comment to remind why we *1e6

File: contracts/SurplusBeneficiary.sol [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        IFeeDistributor(_feeDistributor).burn(token);
+
+        uint256 balanceAfter = IERC20Upgradeable(token).balanceOf(address(this));
+
+        // SB_BNZ: balance is not zero
+        require(balanceAfter == 0, "SB_BNZ");
+
+        emit Dispatch(tokenAmountToTreasury, tokenAmount.sub(tokenAmountToTreasury));
+
+        return tokenAmount;
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:105]: can we remove this if no one need it?

File: contracts/SurplusBeneficiary.sol [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        // transfer to treasury first, because FeeDistributor.burn() will transfer all balance from SurplusBeneficiary
+        SafeERC20Upgradeable.safeTransfer(IERC20Upgradeable(token), _treasury, tokenAmountToTreasury);
+        IFeeDistributor(_feeDistributor).burn(token);
+
+        uint256 balanceAfter = IERC20Upgradeable(token).balanceOf(address(this));
+
+        // SB_BNZ: balance is not zero
+        require(balanceAfter == 0, "SB_BNZ");
+
+        emit Dispatch(tokenAmountToTreasury, tokenAmount.sub(tokenAmountToTreasury));
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:103]: i don't know what's easier for subgraph, but do we also need to record the amount when we can query the amount from ERC20 transfer?
maybe emit target address is more useful, if the target address may be changed by admin

File: test/SurplusBeneficiary/SurplusBeneficiary.spec.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        await testUSDC.__TestERC20_init("USDC", "USDC", 6)
+
+        // use erc20 contract as a treasury contract
+        treasury = await testERC20Factory.deploy()
+
+        const vePERPFactory = await ethers.getContractFactory("vePERP")
+        vePERP = (await vePERPFactory.deploy(testPERP.address, "vePERP", "vePERP", "v1")) as VePERP
+
+        const feeDistributorFactory = await ethers.getContractFactory("FeeDistributor")
+        feeDistributor = (await feeDistributorFactory.deploy(
```
Comment (Wraecca) [inline test/SurplusBeneficiary/SurplusBeneficiary.spec.ts:40]: i think we can just mock a feeDistributor instead of deploying an instance for unit test

File: contracts/SurplusBeneficiary.sol [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+import { IFeeDistributor } from "./interface/IFeeDistributor.sol";
+
+contract SurplusBeneficiary is
+    ISurplusBeneficiary,
+    ReentrancyGuardUpgradeable,
+    OwnableUpgradeable,
+    SurplusBeneficiaryStorageV1
+{
+    using AddressUpgradeable for address;
+    using SafeMathUpgradeable for uint256;
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:26]: do we still need this?

File: contracts/SurplusBeneficiary.sol [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        // SB_TANC: token address is not contract
+        require(tokenArg.isContract(), "SB_TANC");
+        emit TokenChanged(_token, tokenArg);
+        _token = tokenArg;
+    }
+
+    function setFeeDistributor(address feeDistributorArg) public onlyOwner {
+        // SB_FDNC: feeDistributor address is not contract
+        require(feeDistributorArg.isContract(), "SB_FDNC");
+        emit FeeDistributorChanged(_feeDistributor, feeDistributorArg);
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:61]: suggest to add `require feeDistributor.token == _token`

File: contracts/SurplusBeneficiary.sol (hunk 5 above)
Comment (Wraecca) [reply contracts/SurplusBeneficiary.sol:67]: btw if removing isContract should check it's not 0

File: contracts/SurplusBeneficiary.sol (hunk 7 above)
Comment (42bchen) [reply contracts/SurplusBeneficiary.sol:105]: removed

File: contracts/SurplusBeneficiary.sol (hunk 4 above)
Comment (Detoo) [reply contracts/SurplusBeneficiary.sol:54]: it will make some dune analytics queries easier. Some other protocols do this as well (ex. Compound)

File: contracts/SurplusBeneficiary.sol (hunk 5 above)
Comment (Detoo) [reply contracts/SurplusBeneficiary.sol:67]: it does not need to be a contract indeed. Will do both

File: test/SurplusBeneficiary/SurplusBeneficiary.spec.ts (hunk 9 above)
Comment (42bchen) [reply test/SurplusBeneficiary/SurplusBeneficiary.spec.ts:40]: change to use feeDistributor mock contract in spec.ts

File: contracts/SurplusBeneficiary.sol (hunk 10 above)
Comment (42bchen) [reply contracts/SurplusBeneficiary.sol:26]: yes

File: contracts/SurplusBeneficiary.sol (hunk 1 above)
Comment (42bchen) [reply contracts/SurplusBeneficiary.sol:22]: changed

File: contracts/SurplusBeneficiary.sol (hunk 3 above)
Comment (42bchen) [reply contracts/SurplusBeneficiary.sol:74]: fixed

File: contracts/SurplusBeneficiary.sol (hunk 8 above)
Comment (Detoo) [reply contracts/SurplusBeneficiary.sol:103]: a dedicated event is more self-descriptive and more future proof (that we don't need to remember the assumptions we made to the ERC20 transfers in terms of analytics)

File: contracts/SurplusBeneficiary.sol (hunk 10 above)
Comment (Detoo) [reply contracts/SurplusBeneficiary.sol]: yes, we need it to calculate the amount in `Dispatch` event

**Review by Detoo — APPROVED** [review]

**Comment by Detoo** [comment]
@Wraecca we have addressed all requests. Let us know if you have any question

File: contracts/SurplusBeneficiary.sol [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+import { ReentrancyGuard } from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import { SafeERC20, IERC20 } from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { FullMath } from "@uniswap/v3-core/contracts/libraries/FullMath.sol";
+import { SurplusBeneficiaryStorageV1 } from "./storage/SurplusBeneficiaryStorage.sol";
+import { ISurplusBeneficiary } from "./interface/ISurplusBeneficiary.sol";
+import { IFeeDistributor } from "./interface/IFeeDistributor.sol";
+
+contract SurplusBeneficiary is ISurplusBeneficiary, ReentrancyGuard, Ownable, SurplusBeneficiaryStorageV1 {
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:16]: maybe no need to separate storage contract if it's immutable

File: contracts/SurplusBeneficiary.sol [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    ) ReentrancyGuard() Ownable() {
+        // SB_TANC: token is not a contract
+        require(tokenArg.isContract(), "SB_TANC");
+        _token = tokenArg;
+
+        setFeeDistributor(feeDistributorArg);
+        setTreasury(treasuryArg);
+        setTreasuryPercentage(treasuryPercentageArg);
+
+        IERC20(_token).approve(feeDistributorArg, type(uint256).max);
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:34]: dispatch will fail after we setFeeDistribotor to a new one if treasuryPercentage is not 100%
should move to setFeeDistributor or dispatch

File: contracts/interface/ISurplusBeneficiary.sol [hunk 14]
```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.7.6;
+
+interface ISurplusBeneficiary {
+    /// @notice Emitted when dispatch function is trigger
+    /// @param amountToDao Distributed fee amount
+    /// @param amountToFeeDistributor Distributed fee amount
+    event Dispatch(uint256 amountToDao, uint256 amountToFeeDistributor);
```
Comment (Wraecca) [inline contracts/interface/ISurplusBeneficiary.sol:8]: rename amountToDao to amountToTreasury?

File: contracts/SurplusBeneficiary.sol [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+        // SB_TAZ: token amount is zero
+        require(tokenAmount != 0, "SB_TAZ");
+
+        // note the ratio is in `1e6` format
+        uint256 tokenAmountToTreasury = FullMath.mulDiv(tokenAmount, _treasuryPercentage, 1e6);
+
+        // transfer to treasury first, because FeeDistributor.burn() will transfer all balance from SurplusBeneficiary
+        SafeERC20.safeTransfer(IERC20(token), _treasury, tokenAmountToTreasury);
+        IFeeDistributor(_feeDistributor).burn(token);
```
Comment (Wraecca) [inline contracts/SurplusBeneficiary.sol:89]: although it will never return false, but since burn returns boolean, do you think we should add `require(fd.burn())`?

**Review by Wraecca — APPROVED** [review]
looks good! only a few minor suggestions

File: contracts/SurplusBeneficiary.sol (hunk 12 above)
Comment (Detoo) [reply contracts/SurplusBeneficiary.sol:16]: doh! 🤣

File: contracts/SurplusBeneficiary.sol (hunk 13 above)
Comment (Detoo) [reply contracts/SurplusBeneficiary.sol:34]: will do. Nice catch!
